src/metabase/types.clj

```diff
@@ -91,11 +106,55 @@
 (derive :type/Name :type/Category)
 
 
-;;; ------------------------------------------------------------ Util Fns ------------------------------------------------------------
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                                  CUSTOM TYPES                                                  |
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(defsetting ^:private custom-types
```

Comment:
This seems to be overloading settings to include things that are not .... settings. 
Perhaps we should have an alternative mechanism for storing persistent state.
IMO stuff like this should be in a database somewhere rather than stuffed into a map.

Reply:
it is in a database somewhere :)

The Settings table is meant as a catch-all key-value store for everything from user-facing settings to internal information we want to record like this. We've been using Settings this way elsewhere already since I first added the table. We have the concept of "Internal Settings" (by setting the `:internal? true` flag) to mark the ones that are "internal" and not user-facing. 

The Settings code already has tons of stuff like optimizations and ability to store values of different types so it doesn't make sense to duplicate all of that logic just to keep user-facing stuff in a different table from other stuff. Plus separating them out would complicate the story of making Metabase horizontally scalable since we'd have to make sure two tables are in sync everywhere instead of just one

Reply:
Storing a type map in a settings feels really hacky to me. 

We use a relational database for metadata storage, and for this kind of information to live in the same place as say the instance timezone is confusing for no reason. If I were approaching this project from a fresh perspective I'd expect `metabase_field` or something like that to be where I look for this kind of mapping.

Reply:
There's no way a small set of dynamic type mappings would make sense in `metabase_field`, which is a table where each row represents a single Field in a database. If we did try to hack that info into it, we'd have to store an entire type hierarchy for each Field (what if a Field's type is a dynamic type that derives from another dynamic type?). And at best that would end up being something like a `TEXT[]` or a JSON blob itself, similar to what we have now. Iterating over 1,00,000+ fields to look for ones that defined new custom types is not only insane but a massive change in how things work. Of course, that sort of query could be optimized by adding appropriate indexes... but that's not doing much to make things less "hacky".

It could make sense as a separate table entirely, but like I mentioned earlier it seems like premature optimization to put a tiny map of two or three types that's used in exactly one place in it's own table. Only a handful of people have Postgres enums in the first place and until we find a compelling reason to store this sort of stuff in somewhere other than our normal key-value table I don't see a reason to make things overly complicated at this point.

I have already been experimenting with ways to make Metabase horizontally scalable and have been playing with a solution to keep the `settings` table in sync between multiple instances, but adding yet another table we to keep in sync is making the story of horizontal scaling that much more complicated.

Reply:
We already store lots of other stuff in the `setting` table, like `version-info`, which looks like 

```json
{
  "latest": {
    "version": "v0.26.2",
    "released": "2017-09-27T11:09:36.358Z",
    "patch": true,
    "highlights": [
      "Update Redshift Driver",
      "Support Java 9",
      "Fix performance issue with fields listing"
    ]
  },
  "older": [
    {
      "version": "v0.26.1",
      "released": "2017-09-27T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "Fix migration issue on MySQL"
      ]
    },
    {
      "version": "v0.26.0",
      "released": "2017-09-26T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "Segment + Metric X-Rays and Comparisons",
        "Better control over metadata introspection process",
        "Improved Timezone support and bug fixes"
      ]
    },
    {
      "version": "v0.25.2",
      "released": "2017-08-09T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "Bug and performance fixes"
      ]
    },
    {
      "version": "v0.25.1",
      "released": "2017-07-27T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "After upgrading to 0.25, unknown protocol error.",
        "Don't show saved questions in the permissions database lists",
        "Elastic beanstalk upgrades broken in 0.25 "
      ]
    },
    {
      "version": "v0.25.0",
      "released": "2017-07-25T11:09:36.358Z",
      "patch": false,
      "highlights": [
        "Nested questions",
        "Enum and custom remapping support",
        "LDAP authentication support"
      ]
    },
    {
      "version": "v0.24.2",
      "released": "2017-06-01T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "Misc Bug fixes"
      ]
    },
    {
      "version": "v0.24.1",
      "released": "2017-05-10T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "Fix upgrades with MySQL/Mariadb"
      ]
    },
    {
      "version": "v0.24.0",
      "released": "2017-05-10T11:09:36.358Z",
      "patch": false,
      "highlights": [
        "Drill-through + Actions",
        "Result Caching",
        "Presto Driver"
      ]
    },
    {
      "version": "v0.23.1",
      "released": "2017-03-30T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "Filter widgets for SQL Template Variables",
        "Fix spurious startup error",
        "Java 7 startup bug fixed"
      ]
    },
    {
      "version": "v0.23.0",
      "released": "2017-03-21T11:09:36.358Z",
      "patch": false,
      "highlights": [
        "Public links for cards + dashboards",
        "Embedding cards + dashboards in other applications",
        "Encryption of database credentials"
      ]
    },
    {
      "version": "v0.22.2",
      "released": "2017-01-10T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "Fix startup on OpenJDK 7"
      ]
    },
    {
      "version": "v0.22.1",
      "released": "2017-01-10T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "IMPORTANT: Closed a Collections Permissions security hole",
        "Improved startup performance",
        "Bug fixes"
      ]
    },
    {
      "version": "v0.22.0",
      "released": "2017-01-10T11:09:36.358Z",
      "patch": false,
      "highlights": [
        "Collections + Collections Permissions",
        "Multiple Aggregations",
        "Custom Expressions"
      ]
    },
    {
      "version": "v0.21.1",
      "released": "2016-12-08T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "BigQuery bug fixes",
        "Charting bug fixes"
      ]
    },
    {
      "version": "v0.21.0",
      "released": "2016-12-08T11:09:36.358Z",
      "patch": false,
      "highlights": [
        "Google Analytics Driver",
        "Vertica Driver",
        "Better Time + Date Filters"
      ]
    },
    {
      "version": "v0.20.3",
      "released": "2016-10-26T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "Fix H2->MySQL/PostgreSQL migrations, part 2"
      ]
    },
    {
      "version": "v0.20.2",
      "released": "2016-10-25T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "Support Oracle 10+11",
        "Fix H2->MySQL/PostgreSQL migrations",
        "Revision timestamp fix"
      ]
    },
    {
      "version": "v0.20.1",
      "released": "2016-10-18T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "Lots of bug fixes"
      ]
    },
    {
      "version": "v0.20.0",
      "released": "2016-10-11T11:09:36.358Z",
      "patch": false,
      "highlights": [
        "Data access permissions",
        "Oracle Driver",
        "Charting improvements"
      ]
    },
    {
      "version": "v0.19.3",
      "released": "2016-08-12T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "fix Dashboard editing header"
      ]
    },
    {
      "version": "v0.19.2",
      "released": "2016-08-10T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "fix Dashboard chart titles",
        "fix pin map saving"
      ]
    },
    {
      "version": "v0.19.1",
      "released": "2016-08-04T11:09:36.358Z",
      "patch": true,
      "highlights": [
        "fix Dashboard Filter Editing",
        "fix CSV Download of SQL Templates",
        "fix Metabot enabled toggle"
      ]
    },
    {
      "version": "v0.19.0",
      "released": "2016-08-01T21:09:36.358Z",
      "patch": false,
      "highlights": [
        "SSO via Google Accounts",
        "SQL Templates",
        "Better charting controls"
      ]
    },
    {
      "version": "v0.18.1",
      "released": "2016-06-29T21:09:36.358Z",
      "patch": true,
      "highlights": [
        "Fix for Hour of day sorting bug",
        "Fix for Column ordering bug in BigQuery",
        "Fix for Mongo charting bug"
      ]
    },
    {
      "version": "v0.18.0",
      "released": "2016-06-022T21:09:36.358Z",
      "patch": false,
      "highlights": [
        "Dashboard Filters",
        "Crate.IO Support",
        "Checklist for Metabase Admins",
        "Converting Metabase Questions -> SQL"
      ]
    },
    {
      "version": "v0.17.1",
      "released": "2016-05-04T21:09:36.358Z",
      "patch": true,
      "highlights": [
        "Fix for Line chart ordering bug",
        "Fix for Time granularity bugs"
      ]
    },
    {
      "version": "v0.17.0",
      "released": "2016-05-04T21:09:36.358Z",
      "patch": false,
      "highlights": [
        "Tags + Search for Saved Questions",
        "Calculated columns",
        "Faster Syncing of Metadata",
        "Lots of database driver improvements and bug fixes"
      ]
    },
    {
      "version": "v0.16.1",
      "released": "2016-05-04T21:09:36.358Z",
      "patch": true,
      "highlights": [
        "Fixes for several time alignment issues (timezones)",
        "Resolved problem with SQL Server db connections"
      ]
    },
    {
      "version": "v0.16.0",
      "released": "2016-05-04T21:09:36.358Z",
      "patch": false,
      "highlights": [
        "Fullscreen (and fabulous) Dashboards",
        "Say hello to Metabot in Slack"
      ]
    }
  ]
}
```

e.g. we've been using it for generic key-value type storage for a while. I think this makes sense actually because down the road it's easy enough to make our key-value storage backend swappable to something like Redis or Memcached. So right now basically our key-value storage is using a DB backend which puts things in a table named `setting`. I can change the name of the Table if you want to something like `key_value_storage`

Reply:
At any rate we're not really using this Setting for metadata purposes (the types still *are* recorded as part of the appropriate Fields) as we are using it to store state so we can restore it when the server restarts. Since we'll need to keep this state synchronized between multiple instances I think the Settings table makes sense enough in this case. If we were doing actual metadata stuff with it then we could move it somewhere more conducive to that use case but piggybacking off the Settings code dramatically simplifies the persistence story and our multiple-instance sync story rather than complicating it

Reply:
At any rate I think storing simple pieces of state in the `setting` table is currently our best bet since it has the caching optimizations and the like ready to go. I'd definitely have to duplicate a big chunk of that caching optimization code to prevent too many DB calls being made if I tried to implement this PR without using a Setting. And that copied code would need its own tests (the Settings cache is already very well-tested) so we'd be potentially doubling the number of moving parts here which is something I'm wary to do when we already have code set up that will work for our current use-case